torch/csrc/jit/tensorexpr/schedule.cpp

```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
   std::unordered_set<const Var*> func_var_set_;
 };
 
+/**
+ * Actual algorithm for "inlining" of rand().  Bind variables to rand() calls.
+ */
+class RandomInliner : public IRMutator {
+ public:
+  RandomInliner(const std::unordered_set<Function*>& funcs) : funcs_(funcs) {
+    for (auto const* func : funcs_) {
+      func_vars_.insert(func->func_var(0));
+    }
+  }
+
+  static bool is_inner_loop(const For* loop) {
+    Block* body = dynamic_cast<Block*>(loop->body());
+    if (!body) {
+      return false;
+    }
+    if (body->nstmts() != 1) {
+      return false;
+    }
+    return dynamic_cast<For*>(*body->stmts().begin()) == nullptr;
+  }
+
+  Stmt* mutate(const For* v) override {
+    const Var* var = v->var();
+    const Expr* start = v->start();
+    const Expr* stop = v->stop();
+    Stmt* body = v->body();
+    LoopOptions loop_options = v->loop_options();
+
+    // Put the new "let"s in the innermost loop.
+    if (!is_inner_loop(v)) {
+      Stmt* new_body = body->accept_mutator(this);
+      if (new_body == body) {
+        return const_cast<For*>(v);
+      }
+      if (new_body == nullptr) {
+        return nullptr;
+      }
+      return new For(var, start, stop, new_body, loop_options);
+    }
+
+    Stmt* new_body = body->accept_mutator(this);
+    for (auto const& p : random_vars_) {
+      Var* v = p.second;
+      new_body = new LetStmt(v, new Intrinsics(kRand, v->dtype()), new_body);
+    }
+    if (new_body == body) {
+      return const_cast<For*>(v);
+    }
+    if (new_body == nullptr) {
+      return nullptr;
+    }
+    return new For(var, start, stop, new_body, loop_options);
+  }
+
+  // Replace the target variable with the caller expressions.
+  const Expr* mutate(const Var* v) override {
+    auto iter = arg_mapping_.find(v);
+    if (iter == arg_mapping_.end()) {
+      return IRMutator::mutate(v);
+    } else {
+      const Expr* expr = iter->second;
+      // Continue to transform the value from the lookup table.
+      return expr->accept_mutator(this);
+    }
+  }
+
+  Stmt* mutate(const Store* v) override {
+    // Remove buffer writes in the inlined functions.
+    if (func_vars_.count(v->base_handle())) {
+      return nullptr;
+    }
+    return IRMutator::mutate(v);
+  }
+
+  // TODO: Test this with multiple different random functions.
+  //
+  // TODO: Can we have accesses to the same random tensor with different
+  // indices?  We should probably generate different random numbers for those.
```

Comment:
In that case, that particular inlining should be disabled. For example: 

B(i, j) = A(i, j) + A(i+1, j), where A(i, j) = random()

For the correct implementation, B(i, j) and B(i-1, j) have a strong spatial correlation. But if implemented as "rand() + rand()", that spatial correctional goes away.

Reply:
This is a great point, I was assuming nonoverlapping dependences.  I think we probably just want to track the call indices and bail out if there are any mismatches.